[JSC] Use DirectCall for one CallVariant case

Constellation opened the pull request:
#### b704d28c49dd2f680f9ae4a767e0800c3adb7ae0
<pre>
[JSC] Use DirectCall for one CallVariant case
<a href="https://bugs.webkit.org/show_bug.cgi?id=296241">https://bugs.webkit.org/show_bug.cgi?id=296241</a>
<a href="https://rdar.apple.com/156242613">rdar://156242613</a>

Reviewed by Yijia Huang.

Let&apos;s emit DirectCall when we only observe one CallVariant for Call.
Even when it is not inlined, then we should emit DirectCall since it is
always faster than normal Call IC.

* JSTests/stress/stack-overflow-arrity-catch.js: Added.
(shouldBe):
(funcWithManyArgs):
(test):
* LayoutTests/js/script-tests/stack-overflow-arrity-catch.js: Removed.
* LayoutTests/js/stack-overflow-arrity-catch-expected.txt: Removed.
* LayoutTests/js/stack-overflow-arrity-catch.html: Removed.
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleInlining):

Canonical link: <a href="https://commits.webkit.org/297664@main">https://commits.webkit.org/297664@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c66a53dd3a4dc41f93ded98b0f575e5823031663

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112449 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32181 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22658 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/118528 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/62868 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32833 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40744 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85427 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/36167 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115396 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26208 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101171 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65858 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25504 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19305 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62374 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/104963 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95595 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19381 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121854 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/111063 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39523 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29434 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94235 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39905 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97407 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94061 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39306 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17105 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/35584 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18116 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39411 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44899 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/135295 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39047 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36358 "Found 17 new JSC stress test failures: stress/arrowfunction-constructor.js.bytecode-cache, stress/arrowfunction-constructor.js.default, stress/arrowfunction-constructor.js.dfg-eager, stress/arrowfunction-constructor.js.dfg-eager-no-cjit-validate, stress/arrowfunction-constructor.js.eager-jettison-no-cjit, stress/arrowfunction-constructor.js.mini-mode, stress/arrowfunction-constructor.js.no-cjit-collect-continuously, stress/arrowfunction-constructor.js.no-cjit-validate-phases, stress/arrowfunction-constructor.js.no-llint, stress/builtin-function-is-construct-type-none.js.bytecode-cache ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42383 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40789 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
```diff
@@ -2270,9 +2270,38 @@ ByteCodeParser::CallOptimizationResult ByteCodeParser::handleInlining(
+                ASSERT(callNode->op() != TailCallVarargs && callNode->op() != TailCallForwardVarargs);
+                auto emittedCallOp = callNode->op();
+                callNode->convertToDirectCall(m_graph.freeze(executable));
+                return emittedCallOp == TailCall ? CallOptimizationResult::InlinedTerminal : CallOptimizationResult::Inlined;
```
`addCall()` can return a `TailCallInlinedCaller`, which represents the case where a tail call is inlined but the caller still has code to run after the call. Only the `TailCall` opcode represents true terminal control flow (where the function ends immediately after the call). Is my understanding correct?

Constellation replied:
Yes, that's correct. TailCallInlineCaller is still not a terminal.

hyjorc1 reviewed (approved):
r=me

Constellation committed:
[JSC] Use DirectCall for one CallVariant case
https://bugs.webkit.org/show_bug.cgi?id=296241
rdar://156242613

Reviewed by Yijia Huang.

Let's emit DirectCall when we only observe one CallVariant for Call.
Even when it is not inlined, then we should emit DirectCall since it is
always faster than normal Call IC.

* JSTests/stress/stack-overflow-arrity-catch.js: Added.
(shouldBe):
(funcWithManyArgs):
(test):
* LayoutTests/js/script-tests/stack-overflow-arrity-catch.js: Removed.
* LayoutTests/js/stack-overflow-arrity-catch-expected.txt: Removed.
* LayoutTests/js/stack-overflow-arrity-catch.html: Removed.
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleInlining):

Canonical link: https://commits.webkit.org/297664@main
